src/transformers/models/bridgetower/modeling_bridgetower.py

```diff
@@ -0,0 +1,1006 @@
+        else:
+            self.cross_modal_text_transform = nn.ModuleList(
+                [nn.Linear(config.input_text_embed_size, config.hidden_size) for _ in range(config.num_hidden_layers)]
+            )
+            self.cross_modal_image_transform = nn.ModuleList(
+                [nn.Linear(config.input_image_embed_size, config.hidden_size) for _ in range(config.num_hidden_layers)]
+            )
+
+        self.token_type_embeddings = nn.Embedding(2, config.hidden_size)
+
+        self.vit_model = BridgeTowerCLIP(
+            config.vit_embed_dim,
+            config.resolution_before,
+            config.vit_layers,
+            config.vit_width,
+            config.vit_patch_size,
+            config.vit_transformer_width,
+            resolution_after=config.image_size,
+            stop_gradient=config.stop_gradient,
+            vit_layernorm_shared=config.vit_layernorm_shared,
+            vit_remove_last=config.vit_remove_last,
+        )
+
+        roberta_config = RobertaConfig.from_pretrained(config.tokenizer)
+        self.text_transformer = RobertaModel(roberta_config)
+
+        if not config.vit_layernorm_shared and config.vit_layernorm_init_from_vit:
+            for ln in self.vit_model.visual.cross_modal_ln_separate:
+                ln.weight.data = self.vit_model.visual.ln_post.weight.data
+                ln.bias.data = self.vit_model.visual.ln_post.bias.data
+
+        self.cross_modal_image_layers = nn.ModuleList(
+            [BridgeTowerBertCrossLayer(self.tokenizer_config) for _ in range(config.num_hidden_layers)]
+        )
+        self.cross_modal_text_layers = nn.ModuleList(
+            [BridgeTowerBertCrossLayer(self.tokenizer_config) for _ in range(config.num_hidden_layers)]
+        )
+
+        # Class token => Linear => Tanh
+        self.cross_modal_image_pooler = BertPooler(config)
+        self.cross_modal_text_pooler = BertPooler(config)
+
+        # ===================== Initialize BridgeTower Components ===================== #
+        # just for first layer
+        self.cross_modal_text_layernorm = nn.LayerNorm(config.hidden_size)
+        self.cross_modal_image_layernorm = nn.LayerNorm(config.hidden_size)
+
+        if config.link_tower_shared:
+            self.cross_modal_text_link_tower = LinkTower(config)
+            self.cross_modal_image_link_tower = LinkTower(config)
+        else:
+            self.cross_modal_text_link_tower = nn.ModuleList(
+                [LinkTower(config) for _ in range(config.num_hidden_layers - 1)]
+            )
+            self.cross_modal_image_link_tower = nn.ModuleList(
+                [LinkTower(config) for _ in range(config.num_hidden_layers - 1)]
+            )
+
+        self.post_init()
+
+    @add_start_docstrings_to_model_forward(BRIDGETOWER_INPUTS_DOCSTRING)
+    @replace_return_docstrings(output_type=BridgeTowerModelOutput, config_class=_CONFIG_FOR_DOC)
+    def forward(
```

Comment:
As this method is where the bulk of the link tower logic is, it would be good to have a few more comments explaining what's happening. For example - why does `irtr_len > 0` affect the shape fed to `image_link_tower`?

Reply:
I have added comments in the forward() function and simplified/removed some of the irtr logic